src/NServiceBus.RabbitMQ/Receiving/MessageConverter.cs

```diff
@@ -98,50 +97,65 @@ string DefaultMessageIdStrategy(BasicDeliverEventArgs message)
 
         static string ValueToString(object value)
         {
-            var s = value as string;
-            if (s != null)
-            {
-                return s;
-            }
-
             var bytes = value as byte[];
             if (bytes != null)
             {
                 return Encoding.UTF8.GetString(bytes);
             }
 
-            var dictionary = value as IDictionary<string, object>;
+            var dictionary = value as Dictionary<string, object>;
             if (dictionary != null)
             {
-                var valuesToJoin = new List<string>();
+                var sb = new StringBuilder();
 
-                // ReSharper disable once LoopCanBeConvertedToQuery
                 foreach (var kvp in dictionary)
                 {
-                    valuesToJoin.Add(string.Concat(kvp.Key, "=", ValueToString(kvp.Value)));
+                    sb.Append(kvp.Key);
+                    sb.Append("=");
+                    sb.Append(ValueToString(kvp.Value));
+                    sb.Append(",");
                 }
 
-                return string.Join(",", valuesToJoin);
+                if (sb.Length > 0)
+                {
+                    sb.Remove(sb.Length - 1, 1);
+                }
+
+                return sb.ToString();
             }
 
-            var list = value as IList;
+            var list = value as List<object>;
```

Comment:
To some degree, yes, but given the nature of what `ValueToString` is doing I think that is both OK and somewhat required.

We know we're being passed a C# representation of an AMQP field table, so that means there is exactly defined set of possible values here. The way the client represents a field array is with a `List<object>`.